src/coins.cpp

```diff
@@ -269,17 +269,19 @@ bool CCoinsViewCache::Flush() {
 
 bool CCoinsViewCache::Sync()
 {
-    bool fOk = base->BatchWrite(cacheCoins, hashBlock, /*will_erase=*/false);
-    // Instead of clearing `cacheCoins` as we would in Flush(), just clear the
-    // FRESH/DIRTY flags of any coin that isn't spent.
-    for (auto it = cacheCoins.begin(); it != cacheCoins.end(); ) {
+    bool fOk = base->BatchWrite(m_sentinel, hashBlock, /*will_erase=*/false);
+    // Instead of clearing `cacheCoins` as we would in Flush(), just erase the
+    // spent coins. All coins left in the linked list after BatchWrite should be spent.
+    for (auto it{m_sentinel.second.Next()}; it != &m_sentinel; ) {
```

Comment:
Yeah it also means passing in the `cachedCoinsUsage` so we can decrement that when we erase.

Reply:
I ended up doing it. I pass in sentinel and cachedCoinsUsage and delete the spent entries when not erasing.
I think calling erase with the outpoint only is slower than erasing with the iterator itself, so I only erase when calling `Sync`, and clear the map at the end of the loop if `erase=true`.

Reply:
> Hmm, it's not at all obvious that this will be faster, as an erase by outpoint involves hashing and lookup in the hash table. It may be worth benchmarking the two options to make sure.

Sorry, I may not have been clear. I am *not* erasing by outpoint. I only erase the spent coins if `erase=false` by outpoint, because that's the only way to find those in constant time.
Otherwise, for a normal erasing flush, we just call `mapCoins.clear()` at the end to wipe them all at once.
I think I have used the fastest options for each path.

Reply:
Do we want to add more Boost dependencies? I thought we wanted to go the other way.

Reply:
> you still iterate a second time (inside mapCoins.clear()),

Hmm I added this in the latest revision because we not had the mapCoins and could clear it.
Before this patch was just not erasing the entries and removing checking that it is empty in ReallocateCache. If the map is destroyed does it not essentially do a clear as well, so this is a wash whether we clear now or on destruction?